### ISSUE TITLE
[NFT-03M] Incorrect Ownership Change Maintenance

### ISSUE THREAD

**Issue by 0xsambugs** [Mean-Finance/nft-permissions#13]
## NFT-03M: Incorrect Ownership Change Maintenance

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) | <img class="o-severity o-major" src="https://omniscia.io/report-assets/major.png" /> | [NFTPermissions.sol:L157](https://github.com/Mean-Finance/nft-permissions/blob/1e75f88d0da6e93f2afbd1069a8c37553f448638/src/NFTPermissions.sol#L157) |

### Description:

The [`NFTPermissions::_update`](https://github.com/Mean-Finance/nft-permissions/blob/1e75f88d0da6e93f2afbd1069a8c37553f448638/src/NFTPermissions.sol#L153-L163) function will cause an NFT that is burned to retain the permissions it last assigned as the `_assignedPermissions` mapping could still point to valid `AssignedPermissions` structures with non-zero `lastUpdated` entries.

While the present [`NFTPermissions::hasPermission`](https://github.com/Mean-Finance/nft-permissions/blob/1e75f88d0da6e93f2afbd1069a8c37553f448638/src/NFTPermissions.sol#L61-L68) implementation would properly fail in such a case due to the `ERC721::ownerOf` invocation, containing permission entries that would otherwise succeed is an invalid trait that can misbehave in a variety of ways - including the re-creation of a particular NFT that has been burned.

### Impact:

Given that this exhibit represents a potential compromise of the permission system the `NFTPermissions` contract is meant to establish, we consider it of "major" severity.

In the code, we have demonstrated how this particular vulnerability can manifest in a supported case by [EIP-721](https://eips.ethereum.org/EIPS/eip-721) (a burn and consequent mint of the same NFT ID). A case whereby a burned NFT would remain usable for permissions would solely arise if derivative contracts of `NFTPermissions` would implement some form of pre-mint system and override the [`NFTPermissions::hasPermission`](https://github.com/Mean-Finance/nft-permissions/blob/1e75f88d0da6e93f2afbd1069a8c37553f448638/src/NFTPermissions.sol#L61-L68) or `NFTPermissions::ownerOf` function to not fail for un-minted NFTs.

### Example:

```solidity
/// @dev We are overriding this function to update ownership values on transfers
function _update(address _to, uint256 _positionId, address _auth) internal override returns (address _from) {
  _from = super._update(_to, _positionId, _auth);
  if (_to == address(0)) {
    // When token is being burned, we can delete this entry on the mapping
    delete lastOwnershipChange[_positionId];
    _burnCounter++;
  } else if (_from != address(0)) {
    // If the token is being minted, then no there is no need to need to write this
    lastOwnershipChange[_positionId] = block.number;
  }
}
```

### Recommendation:

We advise the code to instead assign the **maximum value possible for a `block.number` in the `NFTPermissions` system** (which would be represented by `type(uint64).max`) to ensure that any previously-assigned permissions before an NFT is burned do not remain active.

Additionally and to avoid this vulnerability manifesting in re-minted NFTs, we advise the `else if` block present in the [`NFTPermissions::_update`](https://github.com/Mean-Finance/nft-permissions/blob/1e75f88d0da6e93f2afbd1069a8c37553f448638/src/NFTPermissions.sol#L153-L163) function to be set to a simple `else` block ensuring that the `lastOwnershipChange` is correctly overwritten from the previously-set maximum.

**Comment by 0xsambugs** [comment]
While this issue is technically true, after we fixed #12, it's no longer possible to call `_mint` directly from a contract that inherits `NFTPermission`. That means that there is no way to re-mint a token that has been burned, since all minting goes through `_mintWithPermissions`, and it assigned a new id each time
Since tokens cannot be re-minted, I would argue that this is no longer an issue

**Comment by 0xsambugs** [comment]
After discussing it with the team, we decided to go ahead and set the max value. This would revoke all previous permissions, regardless of how `ownerOf` is implemented
